packages/core/src/editor/actions.ts

```diff
@@ -21,114 +22,134 @@ import invariant from "tiny-invariant";
 import { deserializeNode } from "../utils/deserializeNode";
 import { createElement } from "react";
 
+// TODO: move to a constants folder
+const editorEmptyState = {
+  nodes: {},
+  events: {
+    dragged: null,
+    selected: null,
+    hovered: null,
+    indicator: null,
+  },
+};
+
 export const Actions = (
   state: EditorState,
   query: QueryCallbacksFor<typeof QueryMethods>
 ) => {
-  const _ = <T extends keyof CallbacksFor<typeof Actions>>(name: T) =>
-    Actions(state, query)[name];
+  /** Helper functions */
+  const addNodeToParentAtIndex = (node: Node, parent: Node, index: number) => {
+    parent.data.nodes.splice(index, 0, node.id);
+    node.data.parent = parent.id;
+    state.nodes[node.id] = node;
+  };
+
+  const getParentAndValidate = (parentId: NodeId): Node => {
+    invariant(parentId, ERROR_NOPARENT);
+    const parent = state.nodes[parentId];
+    invariant(parent, ERROR_INVALID_NODEID);
+    return parent;
+  };
+
   return {
-    setOptions(cb: (options: Partial<Options>) => void) {
-      cb(state.options);
-    },
-    setNodeEvent(eventType: NodeEvents, id: NodeId | null) {
-      const current = state.events[eventType];
-      if (current && id !== current) {
-        state.nodes[current].events[eventType] = false;
+    /**
+     * Add a new Node(s) to the editor.
+     *
+     * @param nodes
+     * @param parentId
+     */
+    add(nodes: Node[] | Node, parentId: NodeId) {
```

Comment:
I'm thinking of making this action simpler - as in to remove the need to support Node[], I initially used it only for `<Canvas />`, but still adding a single Node should suffice.

Perhaps we can remove this and rename `addNodeToIndex` as the new `add`